File: ios-app/LaVeinteDigital/IMSS/Portal/TarjetonDigitalJson.swift
import Foundation

/// Parsing del resultado de `WKWebView.evaluateJavaScript`.
///
/// Acepta tanto un objeto/array directo como un string JSON con un nivel extra
/// (por si algún bridge intermedio lo re-serializa), centralizando el fix del
/// bug de doble serialización (`TarjetonDigitalJson.kt`).
enum TarjetonDigitalJson {

    static func parseObject(_ raw: String?) -> [String: Any]? {
        guard let raw, raw != "null", raw != "undefined" else { return nil }
        if let data = raw.data(using: .utf8),
           let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            return obj
        }
        if let data = raw.data(using: .utf8),
           let str = try? JSONSerialization.jsonObject(with: data) as? String,
           let inner = str.data(using: .utf8),
           let obj = try? JSONSerialization.jsonObject(with: inner) as? [String: Any] {
            return obj
        }
        return nil
    }

    static func parseArray(_ raw: String?) -> [[String: Any]]? {
        guard let raw, raw != "null", raw != "undefined" else { return nil }
        if let data = raw.data(using: .utf8),
           let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
            return arr
        }
        if let data = raw.data(using: .utf8),
           let str = try? JSONSerialization.jsonObject(with: data) as? String,
           let inner = str.data(using: .utf8),
           let arr = try? JSONSerialization.jsonObject(with: inner) as? [[String: Any]] {
            return arr
        }
        return nil
    }

    /// Decodifica un resultado `JSON.stringify("string")` (o una cadena ya JSON).
    static func parseString(_ raw: String?) -> String? {
        guard let raw, raw != "null", raw != "undefined" else { return nil }
        if let data = raw.data(using: .utf8),
           let s = try? JSONSerialization.jsonObject(with: data) as? String {
            return s
        }
        return nil
    }
}

File: android-app/app/src/main/java/com/laveintedigital/app/imss/portal/TarjetonDigitalJson.kt
package com.laveintedigital.app.imss.portal

import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener

/**
 * Parsea el resultado de `WebView.evaluateJavascript`.
 *
 * `evaluateJavascript` serializa el valor devuelto por el script como JSON. Cuando
 * el script hace `JSON.stringify(...)`, el resultado llega con DOBLE serialización:
 *   - objeto:  "{\"a\":1}"   (JSON string que contiene el objeto)
 *   - array:   "[{\"a\":1}]" (JSON string que contiene el array)
 *
 * Estas funciones resuelven ese nivel extra de string de forma centralizada, para
 * no repetir el bug de "leer el JSONArray/JSONObject directamente y obtener null".
 */
object TarjetonDigitalJson {

    fun parseObject(raw: String?): JSONObject? {
        if (raw == null || raw == "null" || raw == "undefined") return null
        return try {
            val v = JSONTokener(raw).nextValue()
            when (v) {
                is JSONObject -> v
                is String -> JSONObject(v)
                else -> null
            }
        } catch (e: Exception) { null }
    }

    fun parseArray(raw: String?): JSONArray? {
        if (raw == null || raw == "null" || raw == "undefined") return null
        return try {
            val v = JSONTokener(raw).nextValue()
            when (v) {
                is JSONArray -> v
                is String -> JSONArray(v)
                else -> null
            }
        } catch (e: Exception) { null }
    }
}
